.circleci/create_circleci_config.py

```diff
@@ -98,11 +102,52 @@ def to_dict(self):
         pytest_flags.append(
             f"--make-reports={self.name}" if "examples" in self.name else f"--make-reports=tests_{self.name}"
         )
+
+        # We need explicit list instead of `pipeline.parameters.tests_to_run` (only available at job runtime)
+        tests = self.tests_to_run
+        if tests is None:
+            folder = os.environ["test_preparation_dir"]
+            test_file = os.path.join(folder, "filtered_test_list.txt")
+            if os.path.exists(test_file):
+                with open(test_file) as f:
+                    tests = f.read().split(" ")
+
+        # expand the test list
+        if tests == ["tests"]:
+            tests = [os.path.join("tests", x) for x in os.listdir("tests")]
+        expanded_tests = []
+        for test in tests:
+            if test.endswith(".py"):
+                expanded_tests.append(test)
+            elif test == "tests/models":
+                expanded_tests.extend([os.path.join(test, x) for x in os.listdir(test)])
+            elif test == "tests/pipelines":
+                expanded_tests.extend([os.path.join(test, x) for x in os.listdir(test)])
+            else:
+                expanded_tests.append(test)
+        # Avoid long tests always being collected together
+        random.shuffle(expanded_tests)
+        tests = " ".join(expanded_tests)
+
+        # Each executor to run ~10 tests
+        n_executors = max(len(tests) // 10, 1)
+        # Avoid empty test list on some executor(s) or launching too many executors
+        if n_executors > self.parallelism:
+            n_executors = self.parallelism
+        job["parallelism"] = n_executors
+
+        # Need to be newline separated for the command `circleci tests split` below
+        command = f'echo {tests} | tr " " "\\n" >> tests.txt'
+        steps.append({"run": {"name": "Get tests", "command": command}})
+
+        command = 'TESTS=$(circleci tests split tests.txt) && echo $TESTS > splitted_tests.txt'
+        steps.append({"run": {"name": "Split tests", "command": command}})
+
+        steps.append({"store_artifacts": {"path": "~/transformers/tests.txt"}})
+        steps.append({"store_artifacts": {"path": "~/transformers/splitted_tests.txt"}})
```

Comment:
Dan we keep the old code if the number of executors is 1? I feel this adds a bit too much complexity.

Reply:
Do you mean using if / else blocks to deal with parallelism 1 and N > 1 separately?